examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+import os
+from collections.abc import Mapping
+from functools import partial
+from typing import Any, cast, List, Literal, Optional, Tuple, Union
+
+import torch
+from lightning_utilities.core import is_overridden
+from lightning_utilities.core.apply_func import apply_to_collection
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from lightning.fabric.fabric import (
+    _PLUGIN_INPUT,
+    _PRECISION_INPUT,
+    _unwrap_objects,
+    Accelerator,
+    Fabric,
+    Logger,
+    Strategy,
+)
+from lightning.fabric.strategies.fsdp import FSDPStrategy
+from lightning.fabric.utilities.rank_zero import rank_zero_warn
+from lightning.fabric.utilities.types import LRScheduler, Optimizable
+from lightning.pytorch.core.module import LightningModule
+
+
+class Trainer:
+    def __init__(
+        self,
+        accelerator: Union[str, Accelerator] = "auto",
+        strategy: Union[str, Strategy] = "auto",
+        devices: Union[List[int], str, int] = "auto",
+        precision: _PRECISION_INPUT = "32-true",
+        plugins: Optional[Union[_PLUGIN_INPUT, List[_PLUGIN_INPUT]]] = None,
+        callbacks: Optional[Union[List[Any], Any]] = None,
+        loggers: Optional[Union[Logger, List[Logger]]] = None,
+        max_epochs: Optional[int] = 1000,
+        max_steps: Optional[int] = None,
+        grad_accum_steps: int = 1,
+        limit_train_batches: Union[int, float] = float("inf"),
+        limit_val_batches: Union[int, float] = float("inf"),
+        val_sanity_check: int = 2,
+        validation_frequency: int = 1,
+        use_distributed_sampler: bool = True,
+        checkpoint_dir: str = "./checkpoints",
+        checkpoint_frequency: int = 1,
+    ) -> None:
+
+        self.fabric = Fabric(
+            accelerator=accelerator,
+            strategy=strategy,
+            devices=devices,
+            precision=precision,
+            plugins=plugins,
+            callbacks=callbacks,
+            loggers=loggers,
+        )
+        self.global_step = 0
+        self.grad_accum_steps: int = grad_accum_steps
+        self.current_epoch = 0
+
+        self.max_epochs = max_epochs
+        self.max_steps = max_steps
+        self.should_stop = False
```

Comment:
Why do we need should stop? Could we use regular breaks or StopIteration instead? should_stop in the Trainer is a bit of a hack

Reply:
I find this one easier to follow, as we cannot easily break multiple loops with `StopIteration`, just the inner-most and then we would at least have a local variable to see if we need to trigger again. 

Also `StopIteration`, in my personal view, is less traceable for users who aren't experienced in python